src/dune_lang/syntax.ml

```diff
@@ -125,23 +182,38 @@ let create ~name ~desc supported_versions =
 
 let name t = t.name
 
-let check_supported t (loc, ver) =
-  if not (Supported_versions.is_supported t.supported_versions ver) then
+let check_supported ~lang_ver t (loc, ver) =
+  if not (Supported_versions.is_supported t.supported_versions ver lang_ver)
+  then
+    let until =
+      match Supported_versions.get_min_lang_ver t.supported_versions ver with
+      | Some v -> Printf.sprintf " until dune lang %s" (Version.to_string v)
+      | None -> ""
+    in
+    let l = Supported_versions.supported_ranges lang_ver t.supported_versions in
+    let supported =
+      ( if List.is_empty l then
+        Pp.textf
+          "There are no supported versions of this extension for dune lang %s."
+      else
+        Pp.textf "Supported versions for dune lang %s:" )
+        (Version.to_string lang_ver)
+    in
+
     User_error.raise ~loc
-      [ Pp.textf "Version %s of %s is not supported." (Version.to_string ver)
-          t.name
-      ; Pp.text "Supported versions:"
-      ; Pp.enumerate (Supported_versions.supported_ranges t.supported_versions)
-          ~f:(fun (a, b) ->
+      [ Pp.textf "Version %s of %s is not supported%s." (Version.to_string ver)
+          t.name until
```

Comment:
```suggestion
          t.desc until
```

The error message will be a bit more precise this way. We'll get: `Version 2.0 of the menhir extension` rather than `Version 2.0 of menhir`.